### PR TITLE
Fix Thermal Comfort zone temperature and humidity lag

### LINKED ISSUE #8516: Fanger thermal comfort calculation misaligned with zone air (and other) temperature values
Issue overview
--------------
In calculating the Fanger thermal comfort values, such as PMV and PPD in a simulation run, the calculated PMV values look like if it is one time-step behind the corresponding temperature values (such as zone air temperature or zone radiant temperature). This is especially the case when the air temperature in a zone abruptly changes (e.g. when the thermostat setpoint changes). 

image

 
The cause of the problem is that the calculation module uses as a saved zone air temperature value for the previous time step, which causes the calculated PMV lagging behind. The calculated PMV value is thus off from what is “supposed” to be, depending on how quickly the zone air temperature changes. 

There is limited impact to the average users who do not need precise real-time thermal comfort information. Currently no workaround is available. The bug is expected to be fixed in future releases. 

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version) All
 - Version of EnergyPlus (if using an intermediate build, include SHA): 9.4, probably all
 - Helpdesk ticket number 15616

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Fixes #8516
 - Changes are made to allow the Fanger PMV and PPD calculation to use up-to-date temperature information, to make accurate (and correct) PMV and PPD calculations. 
 - The following results indicated how the issue was fixed: 
image
 
vs. the original results before the fix: 
image
(See #8516 as well)

  - Diffs: The common types of diffs that have been observed in many cases are the ``uncomfortable hours`` diffs (e.g. based on ASHRAE Standard 55), such as: 
 
image

These diffs could be expected as the PMV values would slightly change after the zone temperature and humidity corrections in the current PR. 

Since the uncomfortable hours are based on computed thermal comfort values (such as PMVs), a graphical comparison of the newly computed PMVs vs. the original PMVs is demonstrated in the following example: 
image
image
image

Additional diff comparisons: 
- Originally reported the psychometric relative humidity out of bound warning increase. After a code change these diffs are now eliminated. 
- No energy use related difference was observed. 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by jcyuan2020** [comment]
New code changes are applied. Hopefully this should solve the increased counts of psychometric RH out of bound warnings.

**Comment by jcyuan2020** [comment]
The psychometric relative humidity ratio out of bound warnings are finally gone--this is good as these warnings are not much expected. 
Now with the changes there are more cases having diffs---comparatively this is not that much unexpected. 
May also due to the fast changing upstream "develop" in the merging flurry--will try some local comparisons for this reason.

**Comment by jmythms** [comment]
For the next reviewer:
I tested out some of the PMV csv diffs on this branch. This branch does better than how the current develop does in generating the PMV value.

For example, looking at the absolute csv diffs in the regression test for RadHiTempElecTermReheat.idf ([CI test results link](http://energyplus.s3-website-us-east-1.amazonaws.com/regressions/2021-03/5f676dbee178472c5a05097e88c8fb5b11644b0f-db8285f0f1546b9f3ebf0996e8edaa2f438bdd84/RadHiTempElecTermReheat/x86_64-MacOS-10.15-clang-11.0.0/)),
the largest PMV diff was 0.818 followed by a diff of 0.747 in the next timestep. 
image

This is how develop performs here:
image

This branch: 
image

As can be observed, the effect of the large jump in temperature on PMV happened one timestep later in the develop branch while it occurred in the same timestep in this branch. Similar big PMV diffs were observed in other regression tests, but could be justified with the temperature changes. 

Looks good!

**Comment by Myoldmopar** [comment]
The Linux failed tests results are unrelated, but the diffs are present.  Would be good to get another set of CI results again.  @jmythms thanks for the results comparison, that helps a lot.  @mjwitte do you have anything further on this before it gets in line for final review?  Anything else you know you'd like to see from @jcyuan2020 ?

File: src/EnergyPlus/HVACManager.cc [hunk 1]
```diff
@@ -243,8 +243,8 @@ namespace EnergyPlus::HVACManager {
 
         state.dataHeatBalFanSys->ZT = state.dataHeatBalFanSys->MAT;
         // save for use with thermal comfort control models (Fang, Pierce, and KSU)
-        state.dataHeatBalFanSys->ZTAVComf = state.dataHeatBalFanSys->ZTAV;
-        state.dataHeatBalFanSys->ZoneAirHumRatAvgComf = state.dataHeatBalFanSys->ZoneAirHumRatAvg;
+        // state.dataHeatBalFanSys->ZTAVComf = state.dataHeatBalFanSys->ZTAV;
+        // state.dataHeatBalFanSys->ZoneAirHumRatAvgComf = state.dataHeatBalFanSys->ZoneAirHumRatAvg;
```
Comment (mjwitte) [inline src/EnergyPlus/HVACManager.cc]: @jcyuan2020 These lines should be deleted instead of commented, but otherwise, I'm good with this.

**Review by mjwitte — APPROVED** [review]

File: src/EnergyPlus/HVACManager.cc (hunk 1 above)
Comment (jcyuan2020) [reply src/EnergyPlus/HVACManager.cc]: @mjwitte The redundant commented code lines are now removed.

File: src/EnergyPlus/HVACManager.cc [hunk 2]
```diff
@@ -491,6 +489,9 @@ void ManageHVAC(EnergyPlusData &state)
         FirstTimeStepSysFlag = false;
     } // system time step  loop (loops once if no downstepping)
 
+    state.dataHeatBalFanSys->ZTAVComf = state.dataHeatBalFanSys->ZTAV;
+    state.dataHeatBalFanSys->ZoneAirHumRatAvgComf = state.dataHeatBalFanSys->ZoneAirHumRatAvg;
+
```
Comment (Myoldmopar) [inline src/EnergyPlus/HVACManager.cc:494]: A small change to make such an impact.  But it seems reasonable, and analysis of the diffs shows better results.  This is fine.

File: tst/EnergyPlus/unit/ThermalComfort.unit.cc [hunk 3]
```diff
@@ -1053,3 +1053,524 @@ TEST_F(EnergyPlusFixture, ThermalComfort_CalcThermalComfortASH55)
...
+    CalcThermalComfortFanger(*state);
+
+    EXPECT_NEAR(state->dataHeatBalFanSys->ZTAVComf(1), 14.863733439268286, 0.001);
+
+    EXPECT_NEAR(state->dataHeatBalFanSys->ZoneAirHumRatAvgComf(1), 0.010564839505489259, 0.0001);
+
+    EXPECT_NEAR(state->dataThermalComforts->ThermalComfortData(1).FangerPMV, -5.5896341565108720, 0.001);
+
+    EXPECT_NEAR(state->dataThermalComforts->ThermalComfortData(1).FangerPPD, 100.0, 0.001);
+}
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ThermalComfort.unit.cc:1576]: OK

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
Pulled in latest develop here and built and ran tests fine.  Code changes look good.  Merging.  Thanks @jcyuan2020 !

**Pull request merged**